scrapy/itemchecker.py

```diff
@@ -0,0 +1,75 @@
+"""
+Filter for Scrapy Items to be used by Feed slots.
+"""
+from itemadapter import ItemAdapter
+
+from scrapy import Item
+from scrapy.utils.misc import load_object
+
+
+class ItemChecker:
+    """
+    This will be used by FeedExporter to decide if an item should be allowed
+    to be exported to a particular feed.
+    :param feed_options: feed specific options passed from FeedExporter
+    :type feed_options: dict
+    """
+
+    item_classes = ()
+
+    def __init__(self, feed_options):
+        self.feed_options = feed_options
+
+        if 'item_classes' in self.feed_options:
+            item_classes = self.feed_options['item_classes']
+
+            if isinstance(item_classes, (Item, str)):
+                self.item_classes += (load_object(item_classes),)
+            elif isinstance(item_classes, (list,set,tuple)):
+                for item_class in item_classes:
+                    self.item_classes += (load_object(item_class),)
+            else:
+                pass
+                # raise some warning for invalid item_classes declaration?
+
+    def accepts(self, item):
+        """
+        Main method to be used by FeedExporter to check if the item is acceptable according
+        to defined constraints. This method uses accepts_class and accept_fields method
+        to decide if the item is acceptable.
+        :param item: scraped item which user wants to check if is acceptable
+        :type item: scrapy supported items (dictionaries, Item objects, dataclass objects, and attrs objects)
+        :return: `True` if accepted, `False` otherwise
+        :rtype: bool
+        """
+        adapter = ItemAdapter(item)
+        return self.accepts_class(item) and self.accepts_fields(adapter.asdict())
```

Comment:
I think here we are introducing a potential, unnecessary performance degradation: we are converting every item into a dict even if `self.accepts_fields` returns `True` for every item. That conversion into a dict may be non-trivial, specially since Scrapy supports arbitrary objects as items by extending https://github.com/scrapy/itemadapter.

Moreover, `self.accept_fields` may want to check field metadata to determine whether or not to accept a field, for which it would need either the original item or the whole adapter.

The only way that I see to solve this performance issue without making `self.accepts_fields` pointless (same signature as `self.accepts_class`) is to actually remove the method.

The more I think about it, the more I see that it’s best to keep this default class simple, to only have the `accepts` method and let it do what `accepts_class` would do. But if you still want to keep the function for some reason, please avoid calling `asdict()` here.